plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataBuilder.java

```diff
@@ -0,0 +1,356 @@
+    this.baseVersion = null;
+    return true;
+  }
+
+  public String getArtifactId() {
+    return artifactId;
+  }
+
+  private boolean setArtifactId(final String artifactId) {
+    checkState(groupId != null, "groupId == null");
+    checkNotNull(artifactId, "artifactId");
+    if (Objects.equals(artifactId, this.getArtifactId())) {
+      return false;
+    }
+    this.artifactId = artifactId;
+    this.baseVersion = null;
+    return true;
+  }
+
+  public String getBaseVersion() {
+    return baseVersion;
+  }
+
+  private boolean setBaseVersion(final String baseVersion) {
+    checkState(groupId != null, "groupId == null");
+    checkState(artifactId != null, "artifactId == null");
+    checkNotNull(baseVersion, "baseVersion");
+    if (Objects.equals(baseVersion, this.getBaseVersion())) {
+      return false;
+    }
+    this.baseVersion = baseVersion;
+    return true;
+  }
+
+  // -----------------------------------
+  // groupId
+
+  public boolean onEnterGroupId(final String groupId) {
+    if (setGroupId(groupId)) {
+      plugins.clear();
+      log.debug("-> G: {}", getGroupId());
+      return true;
+    }
+    return false;
+  }
+
+  @Nullable
+  public Maven2Metadata onExitGroupId() {
+    checkState(getGroupId() != null, "groupId");
+    log.debug("<- G: {}", getGroupId());
+    if (plugins.isEmpty()) {
+      log.debug("No plugins in group: {}", getGroupId());
+      return null;
+    }
+    return Maven2Metadata.newGroupLevel(DateTime.now(), getGroupId(), plugins);
+  }
+
+  public boolean addPlugin(final String prefix, final String artifactId, final String name) {
+    final Plugin plugin = Maven2Metadata.newPlugin(artifactId, prefix, name);
+    final Iterator<Plugin> pi = plugins.iterator();
+    while (pi.hasNext()) {
+      final Plugin p = pi.next();
+      if (plugin.equals(p)) {
+        return false; // bail out, is present already
+      }
+      if (plugin.keyEquals(p)) {
+        pi.remove(); // remove it, will add it below
+        break;
+      }
+    }
+    plugins.add(plugin);
+    log.debug("Added plugin {}:{} prefix:{}", getGroupId(), artifactId, prefix);
+    return true;
+  }
+
+  // -----------------------------------
+  // artifactId
+
+  public boolean onEnterArtifactId(final String artifactId) {
+    if (setArtifactId(artifactId)) {
+      log.debug("-> GA: {}:{}", getGroupId(), getArtifactId());
+      baseVersions.clear();
+      return true;
+    }
+    return false;
+  }
+
+  @Nullable
+  public Maven2Metadata onExitArtifactId() {
+    checkState(getArtifactId() != null, "artifactId");
+    log.debug("<- GA: {}:{}", getGroupId(), getArtifactId());
+    if (baseVersions.isEmpty()) {
+      log.debug("Nothing to generate: {}:{}", getGroupId(), getArtifactId());
+      return null;
+    }
+    Iterator<Version> vi = baseVersions.descendingIterator();
+    String latest = vi.next().toString();
+    String release = latest;
+    while (release.contains("SNAPSHOT") && vi.hasNext()) {
```

Comment:
Do we care about anyone mis-using the SNAPSHOT string in their release versions? (ie. 1.0.SNAPSHOT-2)